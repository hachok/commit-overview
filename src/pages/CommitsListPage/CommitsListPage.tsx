import React, {Fragment, useCallback, useEffect, useState} from 'react';
import {Box, List, makeStyles, TextField, Typography} from '@material-ui/core';
import {useHistory} from 'react-router-dom';
import {
    DateRangePicker,
    DateRangeDelimiter,
    RangeInput, LocalizationProvider
} from '@material-ui/pickers';
import DateFnsUtils from '@material-ui/pickers/adapter/date-fns';
import {format} from 'date-fns';

import Commit from 'src/components/Commit/Commit';
import {API_URL} from 'src/constants';
import {CommitData} from 'src/types';

const useStyles = makeStyles((theme) => ({
    root: {
        width: '100%',
        backgroundColor: theme.palette.background.paper,
    }
}));

const CommitsListPage: React.FC = () => {
    const [commits, setCommits] = useState([]);
    const prevMonthDate = new Date(new Date().setMonth(new Date().getMonth() - 1));
    const [selectedDate, handleDateChange] = React.useState<RangeInput<null>>([prevMonthDate, null]);
    const classes = useStyles();
    const history = useHistory();

    const onCommitClick = (commitId: string) => {
        history.push('/commit/' + commitId);
    };

    const fetchCommitsList = useCallback(async () => {
        const [startDate, endDate] = selectedDate;
        const urlParams = new URLSearchParams();

        if (startDate) urlParams.set('since', format(startDate as Date, 'yyyy-MM-dd'));
        if (endDate) urlParams.set('until', format(endDate as Date, 'yyyy-MM-dd'));

        const urlParamsString = urlParams.toString() ? `?${urlParams.toString()}` : '';

        const res = await fetch(`${API_URL}/commits` + urlParamsString);
        return await res.json();
    }, [selectedDate]);

    useEffect(() => {
        fetchCommitsList().then((commits) => setCommits(commits));
    }, [fetchCommitsList]);

    return (
        <Fragment>
            <Typography variant="h6" color="inherit" noWrap>
                <Box mt={1} mb={2}><strong>facebook / react</strong></Box>
                <LocalizationProvider dateAdapter={DateFnsUtils}>
                    <DateRangePicker
                        startText="Start date"
                        endText="End date"
                        value={selectedDate}
                        onChange={handleDateChange}
                        renderInput={(startProps, endProps) => (
                            <Fragment>
                                <TextField {...startProps} />
                                <DateRangeDelimiter> to </DateRangeDelimiter>
                                <TextField {...endProps} />
                            </Fragment>
                        )}
                    />
                </LocalizationProvider>
            </Typography>
            <Box mt={2} width="100%">
                <List className={classes.root}>
                    {!!commits.length &&
                    commits.map((commit: CommitData) =>
                        <Commit
                            key={commit.sha}
                            info={commit}
                            onClick={onCommitClick}/>)}
                </List>
            </Box>
        </Fragment>);
};

export default CommitsListPage;