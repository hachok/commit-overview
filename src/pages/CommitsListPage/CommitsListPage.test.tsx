import {cleanup, render} from '@testing-library/react';
import CommitsListPage from 'src/pages/CommitsListPage/CommitsListPage';
import '@testing-library/jest-dom/extend-expect'
import fetch from 'jest-fetch-mock';

const mockData = [{
    'sha': 'd74559746c03d9f620e4fd338c70ce319070e0be',
    'commit': {
        'author': {
            'name': 'Ricky',
            'email': 'rickhanlonii@gmail.com',
            'date': '2021-03-07T23:43:29Z'
        },
        'message': 'Use update lane priority to set pending updates on roots (#20918)',
    },
    'author': {
        'avatar_url': 'https://avatars.githubusercontent.com/u/2440089?v=4',
    },
}];

beforeEach(() => {
    fetch.resetMocks();
});

afterEach(cleanup);

test('it should render component correctly', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockData));
    const rendered = render(<CommitsListPage/>);

    await rendered.findByText('Ricky');

    expect(rendered).toMatchSnapshot();
})
