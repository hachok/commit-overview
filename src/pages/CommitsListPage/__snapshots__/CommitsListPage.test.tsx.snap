// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render component correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
      >
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <strong>
            facebook / react
          </strong>
        </div>
        <div
          class="MuiPickersDateRangePickerInput-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
              data-shrink="true"
            >
              Start date
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                placeholder="mm/dd/yyyy"
                type="tel"
                value="02/08/2021"
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                >
                  <span>
                    Start date
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
            >
              mm/dd/yyyy
            </p>
          </div>
          <p
            class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-7 MuiTypography-body1"
          >
             to 
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
              data-shrink="false"
            >
              End date
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input"
                placeholder="mm/dd/yyyy"
                type="tel"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="PrivateNotchedOutline-legendLabelled-5"
                >
                  <span>
                    End date
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-contained"
            >
              mm/dd/yyyy
            </p>
          </div>
        </div>
      </h6>
      <div
        class="MuiBox-root MuiBox-root-8"
      >
        <ul
          class="MuiList-root makeStyles-root-1 MuiList-padding"
        >
          <div
            class="makeStyles-wrapper-9"
            data-testid="commit-element"
          >
            <li
              class="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart"
            >
              <div
                class="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart"
              >
                <div
                  class="MuiAvatar-root MuiAvatar-circle"
                >
                  <img
                    alt="Avatar"
                    class="MuiAvatar-img"
                    src="https://avatars.githubusercontent.com/u/2440089?v=4"
                  />
                </div>
              </div>
              <div
                class="MuiListItemText-root makeStyles-listItem-11 MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Ricky
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  <span
                    class="MuiTypography-root makeStyles-inline-10 MuiTypography-body2 MuiTypography-colorTextPrimary"
                  >
                    08-03-2021
                  </span>
                   - 
                  Use update lane priority to set pending updates on roots (#20918)
                </p>
              </div>
            </li>
            <hr
              class="MuiDivider-root MuiDivider-inset"
            />
          </div>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <h6
      class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
    >
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <strong>
          facebook / react
        </strong>
      </div>
      <div
        class="MuiPickersDateRangePickerInput-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
            data-shrink="true"
          >
            Start date
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              placeholder="mm/dd/yyyy"
              type="tel"
              value="02/08/2021"
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
              >
                <span>
                  Start date
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled"
          >
            mm/dd/yyyy
          </p>
        </div>
        <p
          class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-7 MuiTypography-body1"
        >
           to 
        </p>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
          >
            End date
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              placeholder="mm/dd/yyyy"
              type="tel"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-5"
              >
                <span>
                  End date
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-contained"
          >
            mm/dd/yyyy
          </p>
        </div>
      </div>
    </h6>
    <div
      class="MuiBox-root MuiBox-root-8"
    >
      <ul
        class="MuiList-root makeStyles-root-1 MuiList-padding"
      >
        <div
          class="makeStyles-wrapper-9"
          data-testid="commit-element"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters MuiListItem-alignItemsFlexStart"
          >
            <div
              class="MuiListItemAvatar-root MuiListItemAvatar-alignItemsFlexStart"
            >
              <div
                class="MuiAvatar-root MuiAvatar-circle"
              >
                <img
                  alt="Avatar"
                  class="MuiAvatar-img"
                  src="https://avatars.githubusercontent.com/u/2440089?v=4"
                />
              </div>
            </div>
            <div
              class="MuiListItemText-root makeStyles-listItem-11 MuiListItemText-multiline"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Ricky
              </span>
              <p
                class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                <span
                  class="MuiTypography-root makeStyles-inline-10 MuiTypography-body2 MuiTypography-colorTextPrimary"
                >
                  08-03-2021
                </span>
                 - 
                Use update lane priority to set pending updates on roots (#20918)
              </p>
            </div>
          </li>
          <hr
            class="MuiDivider-root MuiDivider-inset"
          />
        </div>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
